Add blogpost for OptimizedTypedSet

yingsu00 opened the pull request.

aweisberg commented on website/blog/2020-12-04-typedset.md:
```diff
@@ -0,0 +1,53 @@
+1. Resolve hash table probing collision 
+2. Rehash. 
+
+However, both are actually not a problem. In all use cases, whole blocks (instead of several positions of a block) are added to the TypedSet, and the problem of resolving collisions can be resolved by keeping track of the blocks being added. Rehashing is not needed at all since we can know the max number of entries before creating a `TypedSet` for most use cases. 
```
```suggestion
However, `BlockBuilder` is not necessary for either of these issues. In all use cases, whole blocks (instead of several positions of a block) are added to the TypedSet, and the problem of resolving collisions can be resolved by keeping track of the blocks being added. Rehashing is not needed in most cases because we can calculate the maximum number of entries before creating the `TypedSet`. 
```

aweisberg commented on website/blog/2020-12-04-typedset.md:
```diff
@@ -0,0 +1,53 @@
+
+These methods take in a `Block` and remember the positions in the set using `SelectedPositions` objects. The operations can be applied on the same set multiple times, i.e. you can union two Block A and B, then intersect with Block C, then minus(except) Block D. This way the internal operations on the Block elements can be streamlined to more efficient loops. The memory consumption and allocation  ...
+
+## Avoid Computing The Hash Positions Multiple Times
```
```suggestion
## Avoid Recomputing Hash Positions
```

aweisberg commented on website/blog/2020-12-04-typedset.md:
```diff
@@ -0,0 +1,53 @@
+
+## Avoid Computing The Hash Positions Multiple Times
+
+Calculating the hash positions took over 60% of these function costs. For some operations like set intersection, the previous implementation requires creating multiple `TypedSet`s and calculating the hash position multiple times. For example, in `array_intersect` function, it builds one `TypedSet` R, and based on the probe result on R, adds new elements to another `TypedSet` L. 
```
```suggestion
Calculating hash positions took over 60% of original function costs. For some operations like set intersection, the previous implementation requires creating multiple `TypedSet`s and calculating the hash position multiple times. For example, in `array_intersect` function, it builds one `TypedSet` R, and based on the probe result on R, adds new elements to another `TypedSet` L. 
```

aweisberg commented on website/blog/2020-12-04-typedset.md:
```diff
@@ -0,0 +1,53 @@
+
+Calculating the hash positions took over 60% of these function costs. For some operations like set intersection, the previous implementation requires creating multiple `TypedSet`s and calculating the hash position multiple times. For example, in `array_intersect` function, it builds one `TypedSet` R, and based on the probe result on R, adds new elements to another `TypedSet` L. 
+
+![Remote Exchange](/img/blog/2020-12-04-OptimizedTypedSet/TypedSet.png)
```
Contains is capitalized and add isn't

yingsu00 replied:
Replaced them with lower cases everywhere.

aweisberg commented on website/blog/2020-12-04-typedset.md:
```diff
@@ -0,0 +1,53 @@
+
+As shown in the above picture, the “Contains” and “Add” operation would both calculate the hash positions. But actually the hashPosition calculated in hashtable A can be reused by hashtable B if the size and hash functions are the same. This hash position indicates the starting position of the linear probing for both hash tables.
+
+![Remote Exchange](/img/blog/2020-12-04-OptimizedTypedSet/OptimizedTypedSet.png)
```
It's lowercase in both here and should match the diagram above

yingsu00 replied:
Replaced them with lower cases everywhere.

aweisberg commented on website/blog/2020-12-04-typedset.md:
```diff
@@ -0,0 +1,53 @@
+
+## Further reading
+
+For more information please refer to the PR https://github.com/prestodb/presto/pull/15362 
```
```suggestion
For more information please refer to the [PR#15362](https://github.com/prestodb/presto/pull/15362).
```

aweisberg commented on website/blog/2020-12-04-typedset.md:
```diff
@@ -0,0 +1,53 @@
+
+![Remote Exchange](/img/blog/2020-12-04-OptimizedTypedSet/TypedSet.png)
+
+As shown in the above picture, the “Contains” and “Add” operation would both calculate the hash positions. But actually the hashPosition calculated in hashtable A can be reused by hashtable B if the size and hash functions are the same. This hash position indicates the starting position of the linear probing for both hash tables.
```
Why does this switch from L and R to A and B, are these examples showing before and after of the same optimization? I don't really understand the example.

```suggestion
As shown in the above picture, the `Contains` and `Add` operation would both calculate the hash positions. But actually the hashPosition calculated in hash table A can be reused by hash table B if the size and hash functions are the same. This hash position indicates the starting position of the linear probing for both hash tables.
```

yingsu00 replied:
@aweisberg I recreated the picture and rewrote this paragraph like this:

```
As shown in the above picture, the `contains` and `add` operation would both calculate the hash positions for hash table `R`
 and `L` respectively. This hash position indicates the starting position of the linear probing for both hash tables. But actually the
 hashPosition calculated in hash table `R` can be reused by hash table `L` if the hash table sizes and hash functions are the 
same. 

In the new `OptimizedTypedSet`, the first step is the same as the original implementation, in which the right side `Block` is 
added to the set first, and hash table `R` is constructed. Intersecting the set with the left side `Block` was done by creating a 
new hash table `L` of the same size (size precalculated) internally: it checks if the new elements are contained in hash table `R`,
 and if not, add them to hash table `L`. Unlike the original implementation, the `contains` and `add` operation share the same hash position calculation. When the new hash table `L` is constructed, the hash table `R` will be discarded. 
```
Does this sound easier to understand?

aweisberg commented on website/blog/2020-12-04-typedset.md:
```diff
@@ -0,0 +1,53 @@
+
+![Remote Exchange](/img/blog/2020-12-04-OptimizedTypedSet/OptimizedTypedSet.png)
+
+In the new `OptimizedTypedSet`, set intersect operation was done by creating a HashTable of the same size (size preknown) internally, and contains/add operation share the same hash position calculation. When the new Hashtable (A) is constructed, the old one (B) will be discarded. 
```
```suggestion
In the new `OptimizedTypedSet`, set intersect operation was done by creating a hash table of the same size (size precalculated) internally, and contains/add operation share the same hash position calculation. When the new hash table (A) is constructed, the old one (B) will be discarded. 
```

aweisberg commented on website/blog/2020-12-04-typedset.md:
```diff
@@ -0,0 +1,53 @@
+authorFBID: 656599427
+---
+
+Ying Su
```
You can remove this

aweisberg commented on website/blog/2020-12-04-typedset.md:
```diff
@@ -0,0 +1,53 @@
+* intersect
+* except
+
+These methods take in a `Block` and remember the positions in the set using `SelectedPositions` objects. The operations can be applied on the same set multiple times, i.e. you can union two Block A and B, then intersect with Block C, then minus(except) Block D. This way the internal operations on the Block elements can be streamlined to more efficient loops. The memory consumption and allocation  ...
```
This is optimization using `SelectedPositions` is not clear to the average reader.

yingsu00 replied:
> This is optimization using `SelectedPositions` is not clear to the average reader.

@aweisberg Makes sense. I just changed it to 
```
These methods take in a `Block` and remember the selected positions.
```
How does it sound?

yingsu00 commented on website/blog/2020-12-04-typedset.md:
```diff
@@ -0,0 +1,53 @@
+
+Ying Su
+
+Function evaluation is a big part of projection CPU cost. Recently we optimized a set of functions using `TypedSet`, e.g.  `map_concat`, `array_union`, `array_intersect`, and `array_except`. By introducing a new `OptimizedTypeSet`, the above functions were improved for 40% to 80% in CPU and wall time as shown in the JMH benchmarks, and the production queries containing these functions improved fo ...
```
> I'm not sure what `improved for 40% to 80% in CPU` means in terms of the improvement. Is 5x faster or 1.8x faster?
> 
> Some other suggestions

 It's 1.8x faster. I'll change it that way.

yingsu00 replied:
> > I'm not sure what `improved for 40% to 80% in CPU` means in terms of the improvement. Is 5x faster or 1.8x faster?
> > Some other suggestions
> 
> It's 1.8x faster. I'll change it that way.

Actually, I think your rewrite `80% reduction in wall time` is better and clearer so I kept it that way. Thanks for the nice rewrites!

yingsu00 commented:
@aweisberg Ariel, thank you so much for reviewing and rewriting it! I have addressed your comments and rewrote the last part. Will you be able to take another look?

aweisberg reviewed (approved):
LGTM! Thanks for the contribution.

yingsu00 commented:
> LGTM! Thanks for the contribution.

Thank you very much! @aweisberg

yingsu00 commented:
@wenleix Will you be able to merge this? Thank you!

wenleix commented:
Merged #119, thanks for the contribution!
